python/GafferTest/ProcessTest.py

```diff
@@ -0,0 +1,426 @@
+						1 : { 3 : { 4 : { 1 : {} } } },
+						2 : { 3 : { 4 : { 1 : {} } } },
+					}
+				)
+
+			self.assertIn( monitor.plugStatistics( plug ).computeCount, { 5, 6, 8 } )
+
+	@GafferTest.TestRunner.CategorisedTestMethod( { "taskCollaboration" } )
+	def testCollaborationTaskDistribution( self ) :
+
+		# -1 ... -n          plug4
+		#   \ | /
+		#    n+1             plug3
+		#   / | \
+		#  1 ... n           plug2
+		#   \ | /
+		#     0              plug1
+
+		numWorkers = IECore.tbb_global_control.active_value( IECore.tbb_global_control.parameter.max_allowed_parallelism )
+		n = 10000 * numWorkers
+
+		plug1 = Gaffer.Plug()
+		plug2 = Gaffer.Plug()
+		plug3 = Gaffer.Plug()
+		plug4 = Gaffer.Plug()
+		plug1.setInput( plug2 )
+		plug2.setInput( plug3 )
+		plug3.setInput( plug4 )
+
+		dependencies = { -x : {} for x in range( 1, n + 1 ) }
+		dependencies = { x : { n + 1 : dependencies } for x in range( 1, n + 1 ) }
+
+		intPlug = Gaffer.IntPlug()
+		GafferTest.parallelGetValue( intPlug, 100000 ) # Get worker threads running in advance
+
+		with Gaffer.PerformanceMonitor() as monitor, Gaffer.ThreadMonitor() as threadMonitor :
+			GafferTest.runTestProcess( plug1, 0, dependencies )
+
+		self.assertEqual( monitor.plugStatistics( plug1 ).computeCount, 1 )
+		self.assertEqual( monitor.plugStatistics( plug2 ).computeCount, n )
+		self.assertEqual( monitor.plugStatistics( plug3 ).computeCount, 1 )
+		self.assertEqual( monitor.plugStatistics( plug4 ).computeCount, n )
+
+		def assertExpectedThreading( plug, numTasks ) :
+
+			s = threadMonitor.plugStatistics( plug ) # Dict mapping thread ID to number of computes
+			self.assertEqual( sum( s.values() ), numTasks )
+
+			if numTasks == 1 :
+				self.assertEqual( len( s ), 1 )
+			else :
+				# Check that every worker thread contributed some work.
+				self.assertEqual( len( s ), numWorkers )
+				# Check that each worker thread did at least half of its fair
+				# share of work. This assertion is too sensitive in CI so it is
+				# disabled by default. On my local test machine (Dual 16 core
+				# Xeon with hyperthreading) I see pretty reliable success up to
+				# `-threads 32`, and regular failure at `-threads 64` (the
+				# default).
+				if False :
+					for t in s.values() :
+						self.assertGreaterEqual( t, 0.5 * numTasks / numWorkers )
+
+		assertExpectedThreading( plug1, 1 )
+		assertExpectedThreading( plug2, n )
+		assertExpectedThreading( plug3, 1 )
+		assertExpectedThreading( plug4, n )
+
+	@GafferTest.TestRunner.CategorisedTestMethod( { "taskCollaboration:performance" } )
+	@GafferTest.TestRunner.PerformanceTestMethod()
+	def testFanOutGatherPerformance( self ) :
+
+		# Pathological case for cycle checking - huge permutations
+		# of paths through the downstream graph.
+		#
+		#       0
+		#     / | \
+		#    1  2  3
+		#     \ | /
+		#       4
+		#     / | \
+		#    5  6  7
+		#     \ | /
+		#       8
+		#     / | \
+		#    9 10 11
+		#     \ | /
+		#      12        (for width 3 and depth 3)
```

Comment:
> At a basic level, it might help to label the root computation somehow ... since I don't think there is any global convention for whether the root goes at the top or bottom.

I was just following the Gaffer convention, where pretty much all graphs flow top-to-bottom, but I've pushed a fixup containing a comment to make this explicit : https://github.com/GafferHQ/gaffer/pull/5493/commits/91ee00cd510e0054674958e5952a81def81603fa.

> I can't see where testCollaborationTaskDistribution creates a root named "0", or where testFanOutGatherPerformance creates a root named "12" - it seems like the dependency dict starts one level away? 

In all cases, the root process is passed as an argument to `runTestProcess()`.